tools/read-conf.py

```diff
@@ -7,57 +7,105 @@
 import sys
 
 
-def getlist(conf_section, key, fallback=None):
-    string = conf_section.get(key, fallback=fallback)
-    if string:
-        return re.split(r"[, ]+", string)
-    else:
-        return fallback
+def get_string(conf, key):
+    return conf.get(key)
+
+def get_string_array(conf, key):
+    string = get_string(conf, key)
+    return None if string is None else re.split(r"[, ]+", string)
+
+def get_bool(conf, key):
+    return conf.getboolean(key)
+
+def get_int(conf, key):
+    return conf.getint(key)
+
+def get_int_array(conf, key):
+    array = get_string_array(conf, key)
+    return None if array is None else [int(x) for x in array]
+
+
+dtypes = {
+    # section APY
+    "HTML_URL": get_string,
+    "APY_URL": get_string,
+
+    "SUBTITLE": get_string,
+    "SUBTITLE_COLOR": get_string,
+
+    "ALLOWED_LANGS": get_string_array,
+    "ALLOWED_VARIANTS": get_string_array,
+    "ALLOWED_PAIRS": get_string_array,
+
+    "ENABLED_MODES": get_string_array,
+    "DEFAULT_MODE": get_string,
+
+    "TRANSLATION_CHAINING": get_bool,
+
+    "DEFAULT_LOCALE": get_string,
+
+    "SHOW_NAVBAR": get_string,
+
+    "PIWIK_SITEID": get_string,
+    "PIWIK_URL": get_string,
+
+    "LIST_REQUEST_CACHE_EXPIRY": get_int,
+    "LANGUAGE_NAME_CACHE_EXPIRY": get_int,
+    "LOCALIZATION_CACHE_EXPIRY": get_int,
+    "AVAILABLE_LOCALES_CACHE_EXPIRY": get_int,
+
+    # section PERSISTENCE
+    "DEFAULT_EXPIRY_HOURS": get_int,
+
+    # section TRANSLATOR
+    "UPLOAD_FILE_SIZE_LIMIT": get_int,
+
+    "TRANSLATION_LIST_BUTTONS": get_int,
+    "TRANSLATION_LIST_WIDTH": get_int,
+    "TRANSLATION_LIST_ROWS": get_int,
+    "TRANSLATION_LIST_COLUMNS": get_int,
+    "TRANSLATION_LISTS_BUFFER": get_int,
+
+    "INSTANT_TRANSLATION_URL_DELAY": get_int,
+    "INSTANT_TRANSLATION_PUNCTUATION_DELAY": get_int,
+    "INSTANT_TRANSLATION_DELAY": get_int,
+
+    # section UTIL
+    "TEXTAREA_AUTO_RESIZE_MINIMUM_WIDTH": get_int,
+    "BACK_TO_TOP_BUTTON_ACTIVATION_HEIGHT": get_int,
+    "APY_REQUEST_URL_THRESHOLD_LENGTH": get_int,
+    "DEFAULT_DEBOUNCE_DELAY": get_int,
+
+    "INSTALLATION_NOTIFICATION_REQUESTS_BUFFER_LENGTH": get_int,
+    "INSTALLATION_NOTIFICATION_INDIVIDUAL_DURATION_THRESHOLD": get_int,
+    "INSTALLATION_NOTIFICATION_CUMULATIVE_DURATION_THRESHOLD": get_int,
+    "INSTALLATION_NOTIFICATION_DURATION": get_int,
+}
+
 
 def check_config(conf, result):
     # Some error checking:
     for section in conf.sections():
-        if section not in ['APY', 'REPLACEMENTS']:
+        if section not in ['APY', 'PERSISTENCE', 'REPLACEMENTS', 'TRANSLATOR', 'UTIL']:
             raise configparser.Error("\nUnknown section [%s]" % (section,))
 
+    # TODO: either remove or check for all sections
```

Comment:
Instead of checking the result after it's been generated, you could: 1. do `check_config` before you populate `results` and 2. check whether each item in all sections of `config` is present in `dtypes`.

Reply:
More helpful error messages throughout this code would be good, too.

Reply:
I did not write either the test nor the error messages. But I will try to implement your suggestions.